### ISSUE TITLE
pip 9.0.1 fails to process sphinx-autoapi

### ISSUE THREAD

**Issue by pdp10** [rtfd/readthedocs.org#3918]
## Details
I seem to have an error if I use sphinx-autoapi in my docs/requirements.txt file.  
At the moment I have *disabled* the generation of the source code documentation because of this. 

The generation of the source code documentation works fine on my local machine (Linux 64 bit, Python 3.6.5, pip 9.0.3). 

* Project URL: https://github.com/pdp10/sbpipe 
* Build URL (if applicable): #6991372 
* Read the Docs username (if applicable): pdp10

## Expected Result
My source code documentation should be integrated with the rest of my documentation by readthedocs. 

## Actual Result
```
Requirement already satisfied: pip in /home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages (from -r docs/requirements.txt (line 1))
Requirement already satisfied: setuptools in /home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages (from -r docs/requirements.txt (line 2))
Collecting sphinx-autoapi (from -r docs/requirements.txt (line 3))
  Using cached sphinx_autoapi-0.5.0-py2.py3-none-any.whl
Collecting pydocstyle (from sphinx-autoapi->-r docs/requirements.txt (line 3))
  Using cached pydocstyle-2.1.1-py3-none-any.whl
Collecting sphinxcontrib-dotnetdomain (from sphinx-autoapi->-r docs/requirements.txt (line 3))
  Downloading sphinxcontrib-dotnetdomain-0.4.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/setuptools/__init__.py", line 10, in <module>
        from setuptools.extern.six.moves import filter, filterfalse, map
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/setuptools/extern/__init__.py", line 1, in <module>
        from pkg_resources.extern import VendorImporter
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/pkg_resources/__init__.py", line 3017, in <module>
        @_call_aside
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/pkg_resources/__init__.py", line 3003, in _call_aside
        f(*args, **kwargs)
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/pkg_resources/__init__.py", line 3045, in _initialize_master_working_set
        dist.activate(replace=False)
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2577, in activate
        declare_namespace(pkg)
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2151, in declare_namespace
        _handle_ns(packageName, path_item)
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2091, in _handle_ns
        _rebuild_mod_path(path, packageName, module)
      File "/home/docs/checkouts/readthedocs.org/user_builds/sbpipe/envs/latest/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2120, in _rebuild_mod_path
        orig_path.sort(key=position_in_sys_path)
    AttributeError: '_NamespacePath' object has no attribute 'sort'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-01closly/sphinxcontrib-dotnetdomain/
You are using pip version 9.0.1, however version 9.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

**Comment by stsewd** [comment]
I can see your latest builds are passing https://readthedocs.org/projects/sbpipe/builds/, were you able to solve the issue?

**Comment by pdp10** [comment]
No, I didn't solve the issue. I simply:
- commented the Sphinx extension `autoapi-extension` which is required for generating the source code documentation using sphinx-autoapi; 
- removed `sphinx-autoapi` from the requirements.txt used for generating the documentation. 

Also, I noticed that I misspelled the title. It is `sphinx-autoapi`, not `sphinx-autodoc`. Now it's updated.

**Comment by humitos** [comment]
`pip` is installed from the system level version of pip and included in the virtualenv created. Currently we have `9.0.1` at the system level.

I know that we have a `latest` build image that you can configure from YAML (http://docs.readthedocs.io/en/latest/yaml-config.html#build-image) but it seems that when it was created the pip version was still `9.0.1` :(

https://github.com/rtfd/readthedocs-docker-images/blob/master/Dockerfile#L98

So, for the moment there is no possible workaround that I can think of.

This will be automatically fixed when the new release of docker image `3.0` goes out, or a new build of `latest` (or `2.0`, https://github.com/rtfd/readthedocs-docker-images/blob/2.0.3/Dockerfile#L32) image happen.

@agjohnson: I think we can just trigger a rebuild of our images and update the builder servers --not sure how much work is that, though.

**Comment by agjohnson** [comment]
A quick test shows that a new pip doesn't resolve this bug. Is is usually the case, it's actually a setuptools problem. We can pin setuptools to the latest for this project for now. I think you're right that we should probably pin pip/setuptools to match our code in our docker images.

**Comment by agjohnson** [comment]
@pdp10 I've enabled a project flag that pushes the latest setuptools on your project. This should help, let us know how things look with the change to conf.py reverted.

**Comment by pdp10** [comment]
@agjohnson  Thank you for this. 
ReadTheDocs can now build the documentation without raising errors / warnings (see [https://readthedocs.org/projects/sbpipe/builds/7067560/](https://readthedocs.org/projects/sbpipe/builds/7067560/) ).

Said this, the source code documentation is not added to the main documentation, whereas it works fine when this is built locally (see [https://github.com/pdp10/sbpipe/blob/develop/docs/sbpipe.pdf](https://github.com/pdp10/sbpipe/blob/develop/docs/sbpipe.pdf) ).

**Comment by humitos** [comment]
@pdp10 Hi!

I think the original problem regarding pip/setuptools is solved. Also, it should be solved in the general case by #3967. I'm closing this issue.

Regarding the "source code documentation not added", can you open a new issue for this, please? Also, can you be more specific? I mean, those PDFs look very similar to me and I didn't find any difference. What we should look at?

Thanks!

**Comment by pdp10** [comment]
Yes, it works now. I think my browser did not refresh properly when I posted my last comment as the two PDFs were different.

Thank you for your help.
